### DISCUSSION TITLE
Are useMutableSource events automatically batched?

### DISCUSSION THREAD

**Discussion by kirillNevedrov** [pmndrs/zustand#318] [Q&A] [answered]
Hello,

I've read through this issue [https://github.com/pmndrs/zustand/issues/89](https://github.com/pmndrs/zustand/issues/89) to understand how to batch multiple store updates. I see that react concurrent mode will solve that issue, because [react will automatically batch multiple setStates](https://reactjs.org/docs/concurrent-mode-adoption.html#feature-comparison). But I've noticed that [zustand v4](https://github.com/pmndrs/zustand/pull/160) will use useMutableSource instead of useReducer internally. Is it guaranteed that react in concurrent mode will automatically batch multiple update events from useMutableSource?

**Comment by dai-shi** [comment, marked as answer]
1. We don't know yet how uMS behaves when it lands.
2. You can try #160 to see the current behavior.
3. I'd expect the behavior of uMS isn't different from that of useReducer.

That said, I'm not 100% sure how batching behaves in concurrent mode...

**Reply by kirillNevedrov** [reply]
Thank you for the clarification.
